Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,368 @@
+                        self.availableConnections.swapAt(self.availableConnections.startIndex, indexToRemove)
+                        self.availableConnections.removeFirst()
+                    }
+                } else {
+                    connection.release()
+                }
+            }
+        }
+
+        private func makeConnection() -> EventLoopFuture<Connection> {
+            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: self.loopGroup, host: self.key.host, port: self.key.port, configuration: self.configuration) { channel in
+                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }
+            }
+            let address = HTTPClient.resolveAddress(host: self.key.host, port: self.key.port, proxy: self.configuration.proxy)
+            return bootstrap.connect(host: address.host, port: address.port).map { channel in
+                let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                self.configureClose(of: connection)
+                connection.isLeased = true
+                return connection
+            }
+        }
+
+        func release(connection: Connection) {
+            self.lock.withLock {
+                connection.isLeased = false
+                if connection.channel.isActive {
+                    if let firstWaiter = waiters.popFirst() {
+                        firstWaiter.promise.succeed(connection)
+                    } else {
+                        self.availableConnections.append(connection)
+                        leased -= 1
+                    }
+                } else {
+                    if let firstWaiter = waiters.popFirst() {
+                        self.makeConnection().cascade(to: firstWaiter.promise)
+                    } else {
+                        leased -= 1
+                    }
+                }
+            }
+        }
+
+        func closeAllConnections() -> EventLoopFuture<Void> {
+            return self.lock.withLock {
+                let closeFutures = availableConnections.map { $0.channel.close().recover { _ in } }
+                return EventLoopFuture<Void>.andAllSucceed(closeFutures, on: loopGroup.next()).map {
+                    self.availableConnections.removeAll()
+                }
+            }
+        }
+
+        private func closeIfNeeded() {
+            if !self.isClosed.load(), self.leased == 0, self.availableConnections.isEmpty, self.waiters.isEmpty {
+                self.isClosed.store(true)
+                self.parentPool._connectionProviders[key] = nil
+            }
+        }
+
+        struct Waiter {
+            var promise: EventLoopPromise<Connection>
+            var preference: HTTPClient.EventLoopPreference
```

Comment:
are both promise and preference mutable?

Reply:
I don't think they need to, changing this to `let`, thank you.